File: Data Structures & Algorithms/search-for-word/submission-0.cpp
class Solution {
public:
    int dx[4] = {1,-1,0,0};
    int dy[4] = {0,0,-1,1};
    int vis[1001][1001];

    bool isValid(int i,int j,int m, int n){
        return !(i< 0 || j< 0 || i>=m || j >=n);
    }

    bool dfs(int i,int j,vector<vector<char>>& board,string word,string temp){
        if(vis[i][j]==1)
            return false;

        vis[i][j]=1;
        temp+= board[i][j];

        if(word == temp)
            return true;

        int x2,y2;
        bool res=false;
        for(int x=0;x<4;x++){
            x2 = dx[x] + i;
            y2 = dy[x] + j;

            if(isValid(x2,y2,board.size(),board[0].size())){
                res|= dfs(x2,y2,board,word,temp);
            }

            if(res)
                break;
        }

        vis[i][j]=0;
        
        temp.pop_back();

        return res;    
    }

    void visited(vector<vector<char>>& board){
        for(int i=0;i<board.size();i++)
            for(int j=0;j<board[0].size();j++)
                vis[i][j]=0;
    }
    bool exist(vector<vector<char>>& board, string word) {
        for(int i=0;i<board.size();i++)
            for(int j=0;j<board[0].size();j++){
                visited(board);
                if(dfs(i,j,board,word,""))
                    return true;
            }

        return false;    

    }   
};

File: Data Structures & Algorithms/search-for-word-ii/submission-0.java
class Solution {
    int []dx = {1,0,-1,0};
    int []dy = {0,-1,0,1};
    private boolean isValid(int i,int j, int m,int n){
        return !(i<0 || j<0 || i>=m || j>=n);
    }
    private boolean dfs(String word,char[][] board,boolean[][] vis,int i,int j,int ind){
        
        if(ind == word.length()-1)
            return true;
        
        int m = board.length;
        int n = board[0].length;
        vis[i][j]=true;

        int x,y;
        for(int k=0;k<4;k++){
            x = dx[k] + i;
            y = dy[k] + j;

            if(isValid(x,y,m,n)){
                
                if((!vis[x][y])&& word.charAt(ind+1) == board[x][y]){
                    if(dfs(word,board,vis,x,y,ind+1)){
                        vis[i][j]=false;
                        return true;
                    }
                        
                }
            }
        }

        vis[i][j]=false;
        return false;
    }
    public List<String> findWords(char[][] board, String[] words) {
        int n = words.length;
        int x = board.length;
        int y = board[0].length;
        boolean [][]vis= new boolean[x][y];
        boolean found;
        List<String> res= new ArrayList<>();
        for(int i=0;i<n;i++){
            found = false;
            for(int j=0;j<x;j++){
                for(int k=0;k<y;k++){
                    if(board[j][k]==words[i].charAt(0)){
                        if(dfs(words[i],board,vis,j,k,0)){
                            found = true;
                            break;
                        }
                    }
                }

                if(found)
                    break;
            }
            if(found)    
                res.add(words[i]);
        }

        return res;
    }
}
